src/ExtractTileOperations.cpp

```diff
@@ -143,6 +183,138 @@ Tile<3> get_3d_tile_index(const Expr &e) {
+    const Add *dim_expr = add_lhs->b.as<Add>();
+
+    if (!dim_expr) {
+        return {};
+    }
+
+    const Broadcast *base_stride_bc = dim_expr->b.as<Broadcast>();
+
+    if (!base_stride_bc) {
+        return {};
+    }
+
+    int tile_y = base_stride_bc->lanes;
+
+    const Mod *mod = dim_expr->a.as<Mod>();
+
+    if (!mod) {
+        return {};
+    }
+
+    const Broadcast *bc_ramp = mod->a.as<Broadcast>();
+
+    if (!bc_ramp) {
+        return {};
+    }
+
+    int tile_xy = bc_ramp->lanes;
+    int tile_x = tile_xy / tile_y;
+
+    const Ramp *r_ramp = bc_ramp->value.as<Ramp>();
+
+    if (!r_ramp) {
+        return {};
+    }
+
+    int tile_r = r_ramp->lanes;
+
+    // get the base and stride
+    const Ramp *base_stride_ramp = base_stride_bc->value.as<Ramp>();
+
+    if (!base_stride_ramp) {
+        return {};
+    }
+
+    const Broadcast *base_bc = base_stride_ramp->base.as<Broadcast>();
+
+    if (!base_bc) {
+        return {};
+    }
+
+    Expr base = base_bc->value;
+    Expr stride;
+
+    bool found_stride = false;
+
+    // this stride pattern can occur if `tile_r` is the same size as `acc`
+    auto stride_pattern = Broadcast::make(Ramp::make(0, 1, tile_r), tile_x * tile_y) / Broadcast::make((4 / element_width), tile_x * tile_y * tile_r) * Broadcast::make(wild_i32, tile_x * tile_y * tile_r);
+
+    std::vector<Expr> results{};
+    if (expr_match(stride_pattern, add_lhs->a, results)) {
+        found_stride = true;
+        stride = std::move(results[0]);
+    }
+
+    if (!found_stride) {
+        stride_pattern = (Broadcast::make(Ramp::make(0, 1, tile_r), tile_x * tile_y) / Broadcast::make((4 / element_width), tile_x * tile_y * tile_r) + wild_i32) * Broadcast::make(wild_i32, tile_x * tile_y * tile_r);
+        if (expr_match(stride_pattern, add_lhs->a, results)) {
+            found_stride = true;
+            stride = std::move(results[1]);
+            base = std::move(results[0]) * stride + base;
+        }
+    }
+
+    if (!found_stride) {
+        return {};
+    }
+
+    return {true, base, {stride, 0, 0}, {tile_x, tile_y, tile_r}};
+}
+
+struct BaseStride {
+    bool result{false};
+    Expr base{};
+    Expr stride{};
+};
+
+BaseStride get_rhs_tile_index(const Expr &index, int element_width, int tile_x, int tile_y, int tile_r) {
+    const auto rhs_tile2 = get_2d_tile_index(index);
+
+    if (!rhs_tile2.result) {
+        const auto rhs_tile1 = get_1d_tile_index(index);
+
+        if (!rhs_tile1.result) {
+            auto rhs_tile3 = get_3d_rhs_tile_index(index, element_width);
+            if (rhs_tile3.extent[0] != tile_x || rhs_tile3.extent[1] != tile_y || rhs_tile3.extent[2] != tile_r) {
+                return {};
+            }
+
+            return {true, rhs_tile3.base, rhs_tile3.stride[0] * element_width};
+        } else {
+            if (rhs_tile1.extent[0] != tile_y * tile_r) {
+                return {};
+            }
+
+            // times 4 because of the rhs layout
```

Comment:
Could you expand this comment a bit? I assume it's some artifact of how AMX lays things out in memory?

Reply:
This should be addressed by 81bacbb680ef74ab47ec30700a4868b272e8f281.